packages/ia-components/sandbox/language-select/language-select.js

```diff
@@ -0,0 +1,140 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+// Add new flag emoji here
+const flags = {
+  en: '🇬🇧',
+  fr: '🇫🇷',
+  de: '🇩🇪',
+  es: '🇪🇸',
+  hi: '🇮🇳',
+  id: '🇮🇩',
+  it: '🇮🇹',
+  ja: '🇯🇵',
+  mr: '🇮🇳',
+  my: '🇲🇲',
+  pt: '🇵🇹',
+};
+
+/**
+ * Renders the flag emoji and display name for a given language
+ *
+ * @param object props
+ *   @param string name language display name
+ *   @param string value language value
+ *
+ * @returns component stateless
+ */
+const LanguageNode = ({ name, value }) => (
+  <>
+    <span className='flag'>{flags[value]}</span>
+    {name}
+  </>
+);
+
+/**
+ * Renders a selectable language option in the languages dropdown
+ *
+ * @param object props
+ *   @param string name language display name
+ *   @param string value language value
+ *   @param string selectedLanguage selected language value
+ *   @param function onClick click handler from LanguageSelect instance
+ *
+ * @returns component stateless
+ */
+const LanguageOption = ({ name, value, selectedLanguage, onClick }) => (
+  <li className={selectedLanguage === value ? 'selected' : null}>
+    <a onClick={onClick} data-language={value}>
+      <LanguageNode name={name} value={value} />
+    </a>
+  </li>
+);
+
+
+/**
+ * Presentational component
+ * A dropdown to select a language. Clicking/tapping the dropdown trigger
+ * opens a dropdown list of languages that render flag icons and the language
+ * name written in its native language.
+ *
+ * @param array options - array of objects to select from
+ * input object example:
+ * { value: 'en', label: 'English' }
+ * @param function onSelect - event handler
+ * @param string selectedLanguage - language that is selected
+ *
+ * @returns component
+ */
+class LanguageSelect extends React.Component {
+  constructor(props) {
+    const { options, selectedLanguage, onSelect } = props;
+    super(props);
+    this.options = options;
+    this.state = {
+      selectedLanguage,
+      dropdownOpen: false,
+    };
+    this.onSelect = onSelect;
+
+    ['toggleDropdown', 'selectLanguage'].forEach((method) => {
+      this[method] = this[method].bind(this);
```

Comment:
What does this binding do here? Does the `<a onClick={this.toggleDropdown}>` in the `render()` method do the same thing?

Reply:
This is a combination of approaches. The first is to ensure we have instance context within our callbacks. The second is to reduce memory footprint by only creating one bound function in memory rather than one for each option in the dropdown. The `toggleDropdown` method doesn't get attached more than once, but I like to keep the binding in one spot for all DOM methods.

Reply:
For the first approach, when the callback is made from the `onClick`, isn't it always in the context of the current object, `this`?

For the binding multiple times, does it create an instance for every one? I figured each option was just emitting the `onClick` event and all passing it to the same listener since you can distinguish the callers from `e.target`.

Reply:
The context will be the DOM element, not the component instance. The language options' `onClick` handlers are all the same method, whereas having the binding inline like `onClick={this.selectLanguage.bind(this)}` would create a new function in memory for each option.

Reply:
Oh wow that's interesting. Is this a common thing for React? It seems like you'd have to do that everywhere.

Reply:
Yep. They have [a separate article in their docs](https://reactjs.org/docs/handling-events.html) addressing this. This isn't exclusive to React; it's been something to be aware of in most frameworks and libraries.

Reply:
Ahhh. I don't remember having to do that with LitElement, but just found their docs.. it looks like Lit [takes care of that context](https://lit-element.polymer-project.org/guide/events#using-this-in-event-handlers) for you so the context is always the component itself.

Reply:
I just saw in that article that you can do this and avoid the binding (it says experimental so it might not work):
```
class LoggingButton extends React.Component {
  // This syntax ensures `this` is bound within handleClick.
  // Warning: this is *experimental* syntax.
  handleClick = () => {
    console.log('this is:', this);
  }

  render() {
    return (
      <button onClick={this.handleClick}>
        Click me
      </button>
    );
  }
}
```